src/engine/plugins/tile.cpp

```diff
@@ -400,206 +379,197 @@ Status TilePlugin::HandleRequest(const std::shared_ptr<datafacade::BaseDataFacad
         };
 
         std::unordered_map<NodeID, std::vector<SegmentData>> directed_graph;
+        // Reserve enough space for unique edge-based-nodes on every edge.
+        // Only a tile with all unique edges will use this much, but
+        // it saves us a bunch of re-allocations during iteration.
+        directed_graph.reserve(edges.size() * 2);
 
         // Build an adjacency list for all the road segments visible in
         // the tile
         for (const auto &edge : edges)
         {
             if (edge.forward_segment_id.enabled)
             {
-                if (directed_graph.count(edge.u) == 0)
+                // operator[] will construct an empty vector at [edge.u] if there is no value.
+                directed_graph[edge.u].push_back({edge.v, edge.forward_segment_id.id});
+                if (edge_based_node_info.count(edge.forward_segment_id.id) == 0)
                 {
-                    directed_graph[edge.u] = {{edge.v, edge.forward_segment_id.id}};
+                    edge_based_node_info[edge.forward_segment_id.id] = {true,
+                                                                        edge.packed_geometry_id};
                 }
                 else
                 {
-                    directed_graph[edge.u].push_back({edge.v, edge.forward_segment_id.id});
+                    BOOST_ASSERT(
+                        edge_based_node_info[edge.forward_segment_id.id].is_geometry_forward ==
+                        true);
+                    BOOST_ASSERT(
+                        edge_based_node_info[edge.forward_segment_id.id].packed_geometry_id ==
+                        edge.packed_geometry_id);
                 }
-                edge_based_node_info[edge.forward_segment_id.id] = {true, edge.packed_geometry_id};
             }
             if (edge.reverse_segment_id.enabled)
             {
-                if (directed_graph.count(edge.v) == 0)
+                directed_graph[edge.v].push_back({edge.u, edge.reverse_segment_id.id});
+                if (edge_based_node_info.count(edge.reverse_segment_id.id) == 0)
```

Comment:
same here, check for count is unneccessary, if reverse/forward segments are unique. This repeats my question from the last review if these can ever interact?